src/Microsoft.Extensions.Configuration.AzureAppConfiguration/DomainVerifier.cs

```diff
@@ -0,0 +1,145 @@
+// Licensed under the MIT license.
+//
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Security;
+using System.Net.Sockets;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+
+namespace Microsoft.Extensions.Configuration.AzureAppConfiguration
+{
+    /// <summary>
+    /// Verify the domain name of the endpoint matches the certificate.
+    /// </summary>
+    internal static class DomainVerifier
+    {
+        private const int TlsPort = 443;
+        private const string SubjectAltNameOid = "2.5.29.17";
+        private const string CommonName = "CN=";
+        private const string MultiDomainWildcard = "*.";
+
+        public static async Task<string> GetValidDomain(Uri originEndpoint, string srvHostName)
+        {
+            if (string.IsNullOrEmpty(srvHostName))
+            {
+                return string.Empty;
+            }
+
+            IEnumerable<string> validDomains = await GetValidDomainsFromTlsCert(srvHostName).ConfigureAwait(false);
+            
+            foreach (var domain in validDomains)
+            {
+                if (originEndpoint.DnsSafeHost.EndsWith(domain, StringComparison.OrdinalIgnoreCase))
+                {
+                    return domain;
+                }
+            }
+
+            return string.Empty;
+        }
+
+        private static async Task<List<string>> GetValidDomainsFromTlsCert(string endpoint)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(endpoint));
+
+            var validDomains = new List<string>();
+
+            X509Certificate2 cert = null;
+
+            using (var client = new TcpClient(endpoint, TlsPort))
+            using (var sslStream = new SslStream(client.GetStream(), leaveInnerStreamOpen: false))
+            {
+
+                // Initiate the connection, so it will download the server certificate
+                await sslStream.AuthenticateAsClientAsync(endpoint).ConfigureAwait(false);
+
+                var serverCertificate = sslStream.RemoteCertificate;
+
+                if (serverCertificate == null)
+                {
+                    return validDomains;
+                }
+
+                cert = new X509Certificate2(serverCertificate);
+            }
+
+            if (cert == null)
+            {
+                return validDomains;
+            }
+
+            // Get the Subject Alternative Name (SAN) extension
+            var sanExtension = cert.Extensions[SubjectAltNameOid];
+
+            if (sanExtension != null)
+            {
+                IEnumerable<string> formattedSanExtensions = sanExtension
+                    .Format(true)
+                    .Split(new[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
+
+                foreach (string formattedExtension in formattedSanExtensions)
+                {
+                    // Valid pattern should be "DNS Name=*.domain.com"

```

Comment:
Then can you confirm starts with "dns name=" ?

Reply:
It should always start with "dns name=" for DNS names. Update accordingly.